### PR TITLE
MODINREACH-378 ContributionExceptionListener should log the id of the item that caused the exception

### PR DESCRIPTION
**Pull request by pavankumar181**
Resolves - [MODINREACH-378](https://issues.folio.org/browse/MODINREACH-378)

## Purpose
The Purpose of this PR is to improve the Logging.

## Approach
Added Logger statements wherever needed along with parameters itemId, instanceId, CentralServerId and exceptions.

## Evidences
**Created Instance:**
modinnreach-378-evidence

**Updated Holding:**
modinnreach-378-evidence-1

**Deleted Item:**
modinnreach-378-evidence-2

#### TODOS and Open Questions

- [ ] Check logging

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/innreach/domain/listener/KafkaInventoryEventListener.java [hunk 1]
```diff
@@ -43,8 +43,11 @@ public void handleItemEvents(List<ConsumerRecord<String, DomainEvent<Item>>> con
     log.info("Handling inventory item events from Kafka [number of events: {}]", consumerRecords.size());
 
     var events = getEvents(consumerRecords);
-
+    for(DomainEvent<Item> event : events){
+      log.info("handleItemEvents:: Event type: {}, tenant: {}, timestamp: {}, data: {} ", event.getType(), event.getTenant(), event.getTimestamp(), event.getData());
```
Comment (gurleenkaurbp) [inline src/main/java/org/folio/innreach/domain/listener/KafkaInventoryEventListener.java]: event.getData cant give me the ID'd. You will have to print 2 ID's the new and the old one I suppose. But make sure there is no NPE. Make event.getData().getNewEntity().getID() and also  event.getData().getOldEntity().getID().

Comment (pavankumar181) [reply src/main/java/org/folio/innreach/domain/listener/KafkaInventoryEventListener.java]: Added both Id's.

File: src/main/java/org/folio/innreach/domain/listener/KafkaInventoryEventListener.java [hunk 2]
```diff
@@ -73,7 +80,13 @@ public void handleItemEvents(List<ConsumerRecord<String, DomainEvent<Item>>> con
   public void handleInstanceEvents(List<ConsumerRecord<String, DomainEvent<Instance>>> consumerRecords) {
     log.info("Handling inventory instance events from Kafka [number of events: {}]", consumerRecords.size());
     var events = getEvents(consumerRecords);
-
+    for(DomainEvent<Instance> event : events){
+      log.info("handleInstanceEvents:: Event type: {}, tenant: {}, timestamp: {}, data: {} ", event.getType(), event.getTenant(), event.getTimestamp(), event.getData());
+      var oldEntity = event.getData().getOldEntity();
+      var newEntity = event.getData().getNewEntity();
+      if(oldEntity!=null && newEntity!=null)
```
Comment (gurleenkaurbp) [inline src/main/java/org/folio/innreach/domain/listener/KafkaInventoryEventListener.java]: And condition isnt the best thing to do. There can be a case that an oldEntity is null but newEntity is not null incase of new instance/hold/item created

Comment (pavankumar181) [reply src/main/java/org/folio/innreach/domain/listener/KafkaInventoryEventListener.java]: Added new condition.

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java [hunk 3]
```diff
@@ -179,7 +181,7 @@ public void runInstanceContribution(UUID centralServerId, Instance instance) {
     }
 
     runOngoing(centralServerId, (ctx, statistics) -> {
-      log.info("Ongoing: starting ongoing instance contribution job {}", ctx);
+      log.info("Ongoing: starting ongoing instance contribution job {}, centralServerId: {}", ctx, centralServerId);
 
       if (eligibleInstance) {
         log.info("Ongoing: contributing instance");
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: All of the rest of the log statements in this method should have the instance id logged.

**Comment by steveellis** [comment]
Some places where the instance ID should be logged (where I [can't comment](https://github.com/isaacs/github/issues/1655)):
* https://github.com/folio-org/mod-inn-reach/pull/347/files#diff-9d00329d9ed5a1b4b074f097c29e1f73032ca91db29d814dca087d728964f72bR123
* https://github.com/folio-org/mod-inn-reach/pull/347/files#diff-9d00329d9ed5a1b4b074f097c29e1f73032ca91db29d814dca087d728964f72bR179
* All the log statements in the `runItemContribution` method. Log instance ID there and the item id as well in each log.info statement.
* Same for item move method (runItemMove).

Basically I think please just go through the whole class, look at every log.info, and please add the instance and item id to each place where we log something.

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java (hunk 3 above)
Comment (pavankumar181) [reply src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: Added Instance and item ids in every log statement in ContributionJobRunner Class.

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java [hunk 4]
```diff
@@ -472,20 +493,22 @@ private void runOngoing(UUID centralServerId, BiConsumer<ContributionJobContext,
...
     }
     catch (Exception e) {
-      log.info("contributeInstance exception block :{}",e.getMessage());
+      log.info("contributeInstance exception block :",e);
       throw e;
     }
   }
 
   private boolean isUnknownEvent(InstanceIterationEvent event, UUID iterationJobId) {
+    log.info("isUnknownEvent:: parameters event: {}, iterationJobId: {}", event, iterationJobId);
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: This will be very misleading in the logs. It's not that useful to know if a method was called since it already is guranteed to be called. Instead we only want to know if an unknown event occurs. So you should only log it if what is returned is true, and that can be done where `isUnknown` event is called above.

The reason it is misleading is because if you saw this in the logs wouldn't you think that it is an unknown event that was encountered. I would until I looked at the source...

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java [hunk 5]
```diff
@@ -472,20 +493,22 @@ private void runOngoing(UUID centralServerId, BiConsumer<ContributionJobContext,
       endContributionJobContext();
     }
     catch (ServiceSuspendedException | FeignException | InnReachConnectionException | SocketTimeOutExceptionWrapper e) {
-      log.info("exception thrown from runOngoing");
+      log.info("exception thrown from runOngoing :", e);
       throw e;
     }
     catch (Exception e) {
-      log.info("contributeInstance exception block :{}",e.getMessage());
+      log.info("contributeInstance exception block :",e);
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: Need a space after your param.

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java [hunk 6]
```diff
@@ -414,46 +426,55 @@ private void contributeInstance(UUID centralServerId, Instance instance, Statist
...
       itemExceptionListener.logWriteError(e, item.getId());
     } finally {
+      log.info("Initial: deContributeItem finally added processed records and updated stats");
       stats.addRecordsProcessed(1);
       updateStats(stats);
     }
   }
 
   private void runOngoing(UUID centralServerId, BiConsumer<ContributionJobContext, Statistics> processor) {
+    log.info("Initial: runOngoing  centralServerId: {}, processor: {}", centralServerId, processor);
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: Need to be careful here. Does processor have a toString implementation that is meaningful? If not it will simply output the data type which is kind of useless. We probably don't need to log this param unless I'm mistaken.

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java [hunk 7]
```diff
@@ -374,22 +383,25 @@ private void contributeItemsChunk(UUID centralServerId, String bibId, List<Item>
...
       addRecordProcessed();
     } finally {
+      log.info("contributeItemsChunk:: finally added processed records and updated stats");
       stats.addRecordsProcessed(itemsCount);
       statsListener.updateStats(stats);
     }
   }
 
   private void addRecordProcessed() {
+    log.info("addRecordProcessed called");
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: Think this can be removed.

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java [hunk 8]
```diff
@@ -374,22 +383,25 @@ private void contributeItemsChunk(UUID centralServerId, String bibId, List<Item>
       // not possible to guess what item failed when the chunk of multiple items is being contributed
       var recordId = items.size() == 1 ? items.get(0).getId() : null;
       itemExceptionListener.logWriteError(e, recordId);
+      log.info("contributeItemsChunk:: recordId: {}, exception occurred: {}", recordId, e);
       addRecordProcessed();
     } finally {
+      log.info("contributeItemsChunk:: finally added processed records and updated stats");
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: Not sure this is really that meaningful.

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java [hunk 9]
```diff
@@ -351,10 +358,12 @@ private void contributeInstanceItems(UUID centralServerId, Instance instance, St
   }
 
   private void contributeItem(UUID centralServerId, String bibId, Item item, Statistics stats) {
+    log.info("contributeItem:: parameters centralServerId: {}, bibId: {}, item id: {}, stats: {}", centralServerId, bibId, item.getId(), stats);
     contributeItemsChunk(centralServerId, bibId, List.of(item), stats);
   }
 
   private void contributeItemsChunk(UUID centralServerId, String bibId, List<Item> items, Statistics stats) {
+    log.info("contributeItemsChunk:: paramerters centralServerId: {}, bibId: {}, items: {}, stats: {}", centralServerId, bibId, items, stats);
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: Do we want to log the Items list? I doubt that it will be converted to a meaningful string. Even if it is, it will be too much for the logs.

Comment (steveellis) [reply src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: Maybe just log the count of the items list.

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java [hunk 10]
```diff
@@ -318,22 +320,27 @@ public void cancelInitialContribution(UUID contributionId) {
   }
 
   private boolean isEligibleForContribution(UUID centralServerId, Instance instance) {
+    log.info("isEligibleForContribution:: parameters centralServerId: {}, instance id: {}", centralServerId, instance.getId());
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: I think this line and the following log statements on all of these methods that return a boolean is excessive. What we might care about is whether or not the boolean was true, not that the method was called, which it always will be.

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java (hunk 3 above)
Comment (steveellis) [reply src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: I think you may have misunderstood the comment. I'm suggesting that you log the instance ids in the `log.info` statements in this method. A lot of the other logging that you have added is good though and shouldn't be removed unless I have suggestions about it. But here we care about whether these conditions have been hit in this method, for the given instance. This is key to debugging the P1 which is about ongoing contributions.

**Comment by steveellis** [comment]
Please add the ids to each log.info statement in `runItemContribution`, `runInstanceDeContribution`, `runItemContribution`, `runItemDecontribution`. These method seem to not have been included. They should be.

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java [hunk 11]
```diff
@@ -495,16 +518,20 @@ public void completeContribution(ContributionJobContext context) {
   }
 
   private void updateStats(Statistics stats) {
+    log.info("updateStats:: updating statistics stats: {}", stats);
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: This probably doesn't need to be logged.

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java [hunk 12]
```diff
@@ -472,20 +493,22 @@ private void runOngoing(UUID centralServerId, BiConsumer<ContributionJobContext,
...
   private boolean isUnknownEvent(InstanceIterationEvent event, UUID iterationJobId) {
+    log.info("isUnknownEvent:: parameters event: {}, iterationJobId: {}", event, iterationJobId);
     return !Objects.equals(event.getJobId(), iterationJobId);
   }
 
   public void completeContribution(ContributionJobContext context) {
+    log.info("completeContribution:: parameters context: {}", context);
     try {
       contributionService.completeContribution(context.getContributionId());
       log.info("Completed contribution");
```
Comment (steveellis) [inline src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java:511]: should add the contribution id here

**Review by steveellis — APPROVED** [review]

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java (hunk 4 above)
Comment (pavankumar181) [reply src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: Added Boolean log wherever isUnknown event called.

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java (hunk 5 above)
Comment (pavankumar181) [reply src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: Added space after comma.

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java (hunk 6 above)
Comment (pavankumar181) [reply src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: Removed processor parameter because here we need centralServerId which is important in this case.

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java (hunk 7 above)
Comment (pavankumar181) [reply src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: Removed

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java (hunk 8 above)
Comment (pavankumar181) [reply src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: Removed

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java (hunk 9 above)
Comment (pavankumar181) [reply src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: Added items count in the log.

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java (hunk 10 above)
Comment (pavankumar181) [reply src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: Added Boolean log wherever isEligibleForContribution is called.

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java (hunk 11 above)
Comment (pavankumar181) [reply src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: Removed

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java (hunk 12 above)
Comment (pavankumar181) [reply src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java:511]: Added Contribution id in the above log.

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java [hunk 13]
```diff
@@ -322,25 +325,29 @@ private boolean isEligibleForContribution(UUID centralServerId, Instance instanc
...
     return validationService.isEligibleForContribution(centralServerId, item);
   }
 
   private boolean isContributed(UUID centralServerId, Instance instance, Item item) {
+    log.info("isContributed:: parameters centralServerId: {}, instance id: {}, item id: {}", centralServerId, instance.getId(), item.getId());
     return recordContributionService.isContributed(centralServerId, instance, item);
   }
 
   private void contributeInstanceItems(UUID centralServerId, Instance instance, Statistics stats) {
+    log.info("contributeInstanceItems:: parameters centralServerId: {}, instance id: {}, stats: {}", centralServerId, instance.getId(), stats);
```
Comment (gurleenkaurbp) [inline src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: Does Statistics have a toString method ? Will it print the needed detail or just the data type ? Please check otherwise no point in printing this object alone, you might need a toString method to print its internal meaningful information.

Comment (pavankumar181) [reply src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: It does have toString() Method, but do we really need Statistics in this case? If not, can I remove it?

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java [hunk 14]
```diff
@@ -114,12 +115,13 @@ public void runInitialContribution(InstanceIterationEvent event, String topic) {
...
       return;
     }
-    log.info("Initial: processing instance iteration event: {}", event);
+    log.info("Initial: processing instance iteration event instance id: {}", event.getInstanceId());
 
     var instanceId = event.getInstanceId();
-
-    if (isUnknownEvent(event, iterationJobId)) {
-      log.info("Initial: skipping unknown event, current job is {}", iterationJobId);
+    boolean isUnknownEvent = isUnknownEvent(event, iterationJobId);
```
Comment (Vignesh-kalyanasundaram) [inline src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java:121]: As we are having a logger inside the if condition , do we really need to store the value in a variable and log that value ? Moreover , the below line will print like this **Initial: isUnknownEvent: true** or **Initial: isUnknownEvent: false** which will not be useful for us I feel.

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java [hunk 15]
```diff
@@ -133,37 +135,38 @@ public void runInitialContribution(InstanceIterationEvent event, String topic) {
     var centralServerId = context.getCentralServerId();
 
     if (isEligibleForContribution(centralServerId, instance)) {
+      log.info("Initial: ELgible for Contribution centralServerId: {}, instanceId: {}", centralServerId, instanceId);
```
Comment (Vignesh-kalyanasundaram) [inline src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: There is a typo in the word eligible. Please correct it.

Comment (pavankumar181) [reply src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: Corrected

File: src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java (hunk 13 above)
Comment (pavankumar181) [reply src/main/java/org/folio/innreach/batch/contribution/service/ContributionJobRunner.java]: Removed stats parameter in log

**Review by gurleenkaurbp — APPROVED** [review]
